### PR TITLE
Rework the parser

### PR DESCRIPTION
**Pull request by artasparks**
This PR completely reworks the parser, based on https://github.com/Kashomon/bodoko/blob/master/core/sgf_parser.go

- Use a DFA (state-based) single-pass parser rather than a scanner with backtracking. No backtracking is needed now and I find it much easier to reason about.
- Add line/col/index tracking for better error messaging
- Add extensive tests, relying on treepath
- Re-enable treepath test for Apply
- Add a special-case treepath '.' to indicate an empty treepath -- useful for grabbing the root node.

Fixes: 
- https://github.com/otrego/clamshell/issues/83
- https://github.com/otrego/clamshell/issues/21

### CODE REVIEW

File: core/sgf/parser.go [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+
+		switch sd.curstate {
+		case beginningState:
+			err = handleBeginning(sd, pbuf, g)
+			if err != nil {
+				return nil, err
+			}
+
+		case betweenState:
+			err = handleBetween(sd, pbuf)
```
Comment (artasparks) [inline core/sgf/parser.go]:
```
if err := handleBetween(sd, pbuf); err != nil {
  return err
}
```

File: core/game/node.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	// Move indicates a move in the game.
+	Move *Move
+
+	// Properties contain all the raw/unprocessed properties
+	Properties map[string][]string
+
+	// Children of this position.
+	Children []*Node
+
+	//
```
Comment (artasparks) [inline core/game/node.go]: Fix this comment

Comment (artasparks) [reply core/game/node.go]: Fixed, also fixed logic to add parent.

File: core/sgf/parser.go (hunk 1 above)
Comment (artasparks) [reply core/sgf/parser.go]: Done.

File: core/sgf/parser.go [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+//
+//     propData => propData
+//     propData => between
+func handlePropData(stateData *stateData, pbuf *propBuffer) error {
+	if stateData.curchar == rbrace &&
+		// C[foo 1[k\] bar]
+		//           ^
+		// There was a backslash used for escaping a r-brace.
+		stateData.holdChar == backslash {
+		stateData.holdChar = rune(0)
```
Comment (mhoak) [inline core/sgf/parser.go:337]: Is the formatting off in these if-statements? Part of the if statement conditional is in line with the if branch. That seems wrong.

**Review by mhoak — APPROVED** [review]

File: core/sgf/parser.go (hunk 3 above)
Comment (artasparks) [reply core/sgf/parser.go:337]: It's technically correct, but confusing. Fixed

**Pull request merged**